File: src/main/java/com/wave/hextractor/util/SMSChecksumUtils.java
package com.wave.hextractor.util;

import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;

/**
 * The Class SMSChecksumUtils.
 */
public class SMSChecksumUtils {

	/** The Constant SMS_HEADER_START. */
	private static final String SMS_HEADER_START = "TMR SEGA";

	/** The Constant SMS_HEADER_LOCATION. */
	private static final int SMS_HEADER_LOCATION = 0x7FF0;

	/** The Constant SMS_HEADER_SIZE. */
	private static final int SMS_HEADER_SIZE = 0x10;

	/** The Constant SMS_HEADER_COUNTRY_CHECKSUMRANGE_OFFSET. */
	private static final int SMS_HEADER_COUNTRY_CHECKSUMRANGE_OFFSET = 0xF;

	/** The Constant SMS_HEADER_CHECKSUM_OFFSET. */
	private static final int SMS_HEADER_CHECKSUM_OFFSET = 0xA;

	/** The Constant HIGH_NIBBLE_SHIFTS. */
	private static final int HIGH_NIBBLE_SHIFTS = 4;

	/** The Constant SMS_HEADER_COUNTRY_OVERSEAS. */
	private static final int SMS_HEADER_COUNTRY_OVERSEAS = 0x4;

	/** The Constant SMS_CHECKSUM_BYTES. */
	private static final int SMS_CHECKSUM_BYTES = 2;

	/**
	 * Calculated SMS checksum.
	 *
	 * @param file the file
	 * @return the short
	 */
	private static short calculatedSMSChecksum(byte[] file) {
		short checksum = 0;
		for (int i = 0; i < SMS_HEADER_LOCATION; i++) {
			checksum += file[i] & Constants.MASK_8BIT;
		}
		for (int i = SMS_HEADER_LOCATION + SMS_HEADER_SIZE; i < file.length; i++) {
			checksum += file[i] & Constants.MASK_8BIT;
		}
		return checksum;
	}

	/**
	 * Calculate SMS checksum.
	 *
	 * @param file the file
	 * @param updateChecksum the update checksum
	 */
	private static void calculateSMSChecksum(byte[] file, boolean updateChecksum) {
		boolean isSmsOv = isSMSOverseasRom(file);
		Utils.log("Is SMS overseas ROM? " + isSmsOv);
		if (isSmsOv) {
			short checksumRom = getSMSChecksum(file);
			Utils.log(String.format("SMS ROM Checksum: 0x%04X", checksumRom));
			short checksumCalulated = calculatedSMSChecksum(file);
			Utils.log(String.format("SMS SUM Checksum: 0x%04X", checksumCalulated));
			if (checksumRom != checksumCalulated) {
				Utils.log("Different SMS checksums.");
				if (updateChecksum) {
					Utils.log(String.format("Updating SMS ROM checksum to 0x%04X", checksumCalulated));
					byte[] checkSumByte = Utils.shortToBytes(checksumCalulated);
					file[SMS_HEADER_LOCATION + SMS_HEADER_CHECKSUM_OFFSET] = checkSumByte[0];
					file[SMS_HEADER_LOCATION + SMS_HEADER_CHECKSUM_OFFSET + 1] = checkSumByte[1];
				}
			}
		}
	}

	/**
	 * Calculate SMS checksum.
	 *
	 * @param file the file
	 * @param updateChecksum the update checksum
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	private static void calculateSMSChecksum(String file, boolean updateChecksum) throws IOException {
		byte[] fileBytes = Files.readAllBytes(Paths.get(file));
		calculateSMSChecksum(fileBytes, updateChecksum);
		if (updateChecksum) {
			try (FileOutputStream fos = new FileOutputStream(file)) {
				fos.write(fileBytes);
			}
		}
	}

	/**
	 * Check SMS checksum.
	 *
	 * @param file the file
	 */
	public static void checkSMSChecksum(byte[] file) {
		calculateSMSChecksum(file, false);
	}

	/**
	 * Check SMS checksum.
	 *
	 * @param file the file
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	public static void checkSMSChecksum(String file) throws IOException {
		calculateSMSChecksum(file, false);
	}

	/**
	 * Check update SMS checksum.
	 *
	 * @param file the file
	 */
	public static void checkUpdateSMSChecksum(byte[] file) {
		calculateSMSChecksum(file, true);
	}

	/**
	 * Check update SMS checksum.
	 *
	 * @param file the file
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	public static void checkUpdateSMSChecksum(String file) throws IOException {
		calculateSMSChecksum(file, true);
	}

	/**
	 * Gets the SMS checksum.
	 *
	 * @param file the file
	 * @return the SMS checksum
	 */
	private static short getSMSChecksum(byte[] file) {
		return Short.reverseBytes(
				ByteBuffer.wrap(getSMSHeader(file), SMS_HEADER_CHECKSUM_OFFSET, SMS_CHECKSUM_BYTES).getShort());
	}

	/**
	 * Gets the SMS header.
	 *
	 * @param file the file
	 * @return the SMS header
	 */
	private static byte[] getSMSHeader(byte[] file) {
		return Arrays.copyOfRange(file, SMS_HEADER_LOCATION, SMS_HEADER_LOCATION + SMS_HEADER_SIZE);
	}

	/**
	 * Checks if is SMS overseas rom.
	 *
	 * @param file the file
	 * @return true, if is SMS overseas rom
	 */
	private static boolean isSMSOverseasRom(byte[] file) {
		byte[] header = getSMSHeader(file);
		return header[SMS_HEADER_COUNTRY_CHECKSUMRANGE_OFFSET] >> HIGH_NIBBLE_SHIFTS == SMS_HEADER_COUNTRY_OVERSEAS
				&& new String(header).startsWith(SMS_HEADER_START);
	}

	/**
	 * Instantiates a new SMS checksum utils.
	 */
	private SMSChecksumUtils() {
	}
}

File: src/main/kotlin/com/wave/hextractor/util/SMSChecksumUtils.kt
package com.wave.hextractor.util

import java.io.FileOutputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

/**
 * The Class SMSChecksumUtils.
 */
object SMSChecksumUtils {
    private const val SMS_HEADER_START = "TMR SEGA"
    private const val SMS_HEADER_LOCATION = 0x7FF0
    private const val SMS_HEADER_SIZE = 0x10
    private const val SMS_HEADER_COUNTRY_CHECKSUM_RANGE_OFFSET = 0xF
    private const val SMS_HEADER_CHECKSUM_OFFSET = 0xA
    private const val HIGH_NIBBLE_SHIFTS = 4
    private const val SMS_HEADER_COUNTRY_OVERSEAS = 0x4
    private const val SMS_CHECKSUM_BYTES = 2

    private fun calculatedSMSChecksum(file: ByteArray): Short {
        var checksum: Short = 0
        for (i in 0 until SMS_HEADER_LOCATION) {
            checksum = (checksum + (file[i].toInt() and Constants.MASK_8BIT).toShort()).toShort()
        }
        for (i in SMS_HEADER_LOCATION + SMS_HEADER_SIZE until file.size) {
            checksum = (checksum + (file[i].toInt() and Constants.MASK_8BIT).toShort()).toShort()
        }
        return checksum
    }

    /**
     * Calculate SMS checksum.
     *
     * @param file the file
     * @param updateChecksum the update checksum
     */
    private fun calculateSMSChecksum(file: ByteArray, updateChecksum: Boolean) {
        val isSmsOv = isSMSOverseasRom(file)
        Utils.log("Is SMS overseas ROM? $isSmsOv")
        if (isSmsOv) {
            val checksumRom = getSMSChecksum(file)
            Utils.log(String.format("SMS ROM Checksum: 0x%04X", checksumRom))
            val checksumCalculated = calculatedSMSChecksum(file)
            Utils.log(String.format("SMS SUM Checksum: 0x%04X", checksumCalculated))
            if (checksumRom != checksumCalculated) {
                Utils.log("Different SMS checksums.")
                if (updateChecksum) {
                    Utils.log(String.format("Updating SMS ROM checksum to 0x%04X", checksumCalculated))
                    val checkSumByte = Utils.shortToBytes(checksumCalculated)
                    file[SMS_HEADER_LOCATION + SMS_HEADER_CHECKSUM_OFFSET] =
                        checkSumByte[0]
                    file[SMS_HEADER_LOCATION + SMS_HEADER_CHECKSUM_OFFSET + 1] =
                        checkSumByte[1]
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun calculateSMSChecksum(file: String, updateChecksum: Boolean) {
        val fileBytes = Files.readAllBytes(Paths.get(file))
        calculateSMSChecksum(fileBytes, updateChecksum)
        if (updateChecksum) {
            FileOutputStream(file).use { fos -> fos.write(fileBytes) }
        }
    }

    /**
     * Check update SMS checksum.
     *
     * @param file the file
     * @throws IOException Signals that an I/O exception has occurred.
     */
    @Throws(IOException::class)
    fun checkUpdateSMSChecksum(file: String) {
        calculateSMSChecksum(file, true)
    }

    private fun getSMSChecksum(file: ByteArray): Short {
        return java.lang.Short.reverseBytes(
            ByteBuffer.wrap(getSMSHeader(file), SMS_HEADER_CHECKSUM_OFFSET, SMS_CHECKSUM_BYTES).short
        )
    }

    private fun getSMSHeader(file: ByteArray): ByteArray {
        return file.copyOfRange(SMS_HEADER_LOCATION, SMS_HEADER_LOCATION + SMS_HEADER_SIZE)
    }

    private fun isSMSOverseasRom(file: ByteArray): Boolean {
        val header = getSMSHeader(file)
        return (
            header[SMS_HEADER_COUNTRY_CHECKSUM_RANGE_OFFSET].toInt() shr HIGH_NIBBLE_SHIFTS ==
                SMS_HEADER_COUNTRY_OVERSEAS &&
                String(header).startsWith(SMS_HEADER_START)
            )
    }
}
